File: DS/Linear-ds/Dynamic-ds/Queue/queue.go
// Implement queue using slice
package main

import "fmt"

type Queue struct {
	item []int
}

// Perfom enqueue (insertion)
func (q *Queue) Enqueue(item int) {
	q.item = append(q.item, item)
}

// Perfrom dequeue (deletion)
func (q *Queue) Dequeue() int {
	// enter item on first posititon
	item := q.item[0]

	// Remove 0th element and create a new slice of remaining elements
	q.item = q.item[1:]

	// returnt the item
	return item
}

// Check if empty
func (q * Queue) isEmpty() bool {
	return len(q.item) == 0
}

func main(){
	q := &Queue{}

	// Enqueue
	q.Enqueue(1)
	q.Enqueue(2)
	q.Enqueue(3)
	q.Enqueue(4)

	// Dequeue
	fmt.Println(q.Dequeue())
}

File: DS/Linear-ds/Dynamic-ds/Queue/Queue.js
// TODO: implement the queue ds
/*
function provided  by the queue ds :
    1. enqueue: add elements at end of the queue
    2. dequeue: remove elements from the front of the queue
    3. peek: get the front element without removing it
    4. isEmpty: check whether an element is present within an queue
    5. printQueue: print the elements present in queue
*/

function Queue() {
    collection = [];
    this.print = function () {
        console.log(collection);
    };

    // Add element to the end of the queue
    this.enqueue = function (element) {
        collection.push(element)
    }

    // Remove elements from front
    this.dequeue = function () {
        return collection.shift()
    }

    // Return the element in front
    this.front = function () {
        return collection[0]
    }

    // Return size of queue
    this.size = function () {
        return collection.length
    }

    // check if collection is empty
    this.isEmpty = function () {
        return (collection.length === 0)
    }
}

let obj = new Queue()

obj.enqueue('Awesome') // add Awesome to collection
obj.enqueue('Javascript') // add Javascript to collection
obj.print() // output: ['Awesome', 'Javascript']
obj.dequeue() // remove element form front -> i.e Awesome
obj.print() // output: ['Javascript']
console.log(obj.front())  // output: Javascript